tests/DCPS/UnitTests/DynamicData/DynamicData.cpp

```diff
@@ -0,0 +1,905 @@
+#else
+  EXPECT_STREQ((const char*)a.ld, (const char*)b.ld);
+#endif
+}
+
+TEST(Mutable, ReadValueFromStruct)
+{
+  const XTypes::TypeIdentifier& ti = DCPS::getCompleteTypeIdentifier<DCPS::SingleValueStruct_xtag>();
+  const XTypes::TypeMap& type_map = DCPS::getCompleteTypeMap<DCPS::SingleValueStruct_xtag>();
+  const XTypes::TypeMap::const_iterator it = type_map.find(ti);
+  EXPECT_TRUE(it != type_map.end());
+
+  XTypes::TypeLookupService tls;
+  tls.add(type_map.begin(), type_map.end());
+  XTypes::DynamicType_rch dt = tls.complete_to_dynamic(it->second.complete, DCPS::GUID_t());
+
+  unsigned char single_value_struct[] = {
+    0x00,0x00,0x00,0xc2, // +4=4 dheader
+    0x20,0x00,0x00,0x00, 0x00,0x00,0x00,0x03, // +4+4=12 my_enum
+    0x20,0x00,0x00,0x01, 0x00,0x00,0x00,0x0a, // +4+4=20 int_32
+    0x20,0x00,0x00,0x02, 0x00,0x00,0x00,0x0b, // +4+4=28 uint_32
+    0x00,0x00,0x00,0x03, 0x05, (0), (0), (0), // +4+1+(3)=36 int_8
+    0x00,0x00,0x00,0x04, 0x06, (0), (0), (0), // +4+1+(3)=44 uint_8
+    0x10,0x00,0x00,0x05, 0x11,0x11, (0), (0), // +4+2+(2)=52 int_16
+    0x10,0x00,0x00,0x06, 0x22,0x22, (0), (0), // +4+2+(2)=60 uint_16
+    0x30,0x00,0x00,0x07, 0x7f,0xff,0xff,0xff,0xff,0xff,0xff,0xff, // +4+8=72 int_64
+    0x30,0x00,0x00,0x08, 0xff,0xff,0xff,0xff,0xff,0xff,0xff,0xff, // +4+8=84 uint_64
+    0x20,0x00,0x00,0x09, 0x3f,0x80,0x00,0x00, // +4+4=92 float_32
+    0x30,0x00,0x00,0x0a, 0x3f,0xf0,0x00,0x00,0x00,0x00,0x00,0x00, // +4+8=104 float_64
+    0x40,0x00,0x00,0x0b, 0x00,0x00,0x00,0x10,
+    0x3f,0xff,0,0,0,0,0,0,0,0,0,0,0,0,0,0,    // +4+4+16=128 float_128
+    0x00,0x00,0x00,0x0c, 'a', (0), (0), (0),  // +4+1+(3)=136 char_8
+    0x10,0x00,0x00,0x0d, 0x00,0x61, (0), (0), // +4+2+(2)=144 char_16
+    0x00,0x00,0x00,0x0e, 0xff, (0), (0), (0), // +4+1+(3)=152 byte
+    0x00,0x00,0x00,0x0f, 0x01, (0), (0), (0), // +4+1+(3)=160 bool
+    0x20,0x00,0x00,0x10, 0x00,0x00,0x00,0x0c, // +4+4=168 nested_struct
+    0x30,0x00,0x00,0x11, 0x00,0x00,0x00,0x04, 'a','b','c','\0', // +4+8=180 str
+    0x40,0x00,0x00,0x12, 0x00,0x00,0x00,0x0a,
+    0x00,0x00,0x00,0x06, 0,0x61,0,0x62,0,0x63 // +4+4+10=198 swtr
+  };
+
+  SingleValueStruct expected;
+  expected.my_enum = E_UINT8; expected.int_32 = 10; expected.uint_32 = 11;
+  expected.int_8 = 5; expected.uint_8 = 6; expected.int_16 = 0x1111;
+  expected.uint_16 = 0x2222; expected.int_64 = 0x7fffffffffffffff;
+  expected.uint_64 = 0xffffffffffffffff; expected.float_32 = 1.0f;
+  expected.float_64 = 1.0; set_float128_value(expected.float_128);
+  expected.char_8 = 'a'; expected.char_16 = 0x0061; expected.byte = 0xff;
+  expected._cxx_bool = true; expected.nested_struct.l = 12;
+  expected.str = "abc"; expected.wstr = L"abc";
+
+  ACE_Message_Block msg(1024);
+  msg.copy((const char*)single_value_struct, sizeof(single_value_struct));
+  XTypes::DynamicData data(&msg, xcdr2, dt);
+
+  // For reading value indirectly through a nested DynamicData object.
+  XTypes::MemberId random_id = 111;
```

Comment:
Is there something special about 111?

Reply:
No, it's just an arbitrary ID that is used to test `DynamicData` that has data for non-constructed types.